### DISCUSSION TITLE
Support for Livolo blind/cover controller

### DISCUSSION THREAD

**Discussion by yonaesp** [Koenkk/zigbee2mqtt#8525] [Devices]
Hello, I have a new livolo controller for blinds, but from what I see it is not implemented yet, maybe it is very new yet. Product link: https://es.aliexpress.com/item/1005001314281332.html
Product Code: VL-C702WZ
I have seen that there is support for a cover controller, I do not know if perhaps the support of this device can serve to “cheat” and be able to use it with this controller.

I managed to add it after multiple attempts, but it is detected as a double switch, although the on / off buttons from zigbee2mqtt’s dhcp panel do not work.

When I added -cover to the end of the model by editing the database, it appeared as a cover controller, but it was not able to connect, maybe it is not the correct way to modify the model. if i can try something i’m all ears, thanks!

**Comment by Piscator74** [comment]
I have the Livolo curtain switch partly working
* It connects as a curtain switch
* Zigbee2mqtt can read out the status of the two switches (left -ON/OFF and right ON/OFF).

However, I cannot control the curtain switch yet. And that will take quite some time to accomplish.
You can read up on it here and how to do it: https://github.com/Koenkk/zigbee2mqtt/issues/5516#issuecomment-753620497

If you are interested, I can share what I did to make it work this far.

**Reply by yonaesp** [reply]
hello, I also managed to detect it as a double switch, but it did not work, I would appreciate if you share how you have done to make it work, regards.

**Comment by Piscator74** [comment]
I have added the curtain switch in the following pull request: https://github.com/Koenkk/zigbee-herdsman-converters/pull/3339

Make sure to first force remove your curtain switch from your zigbee network and pair it after you apply the changes from the pull request. Otherwise, it will still be recognized as the incorrect switch (double switch). The log should explicitly say curtain switch.

Good luck! Let me know if it works for you.

**Reply by yonaesp** [reply]
> I have added the curtain switch in the following pull request: [Koenkk/zigbee-herdsman-converters#3339](https://github.com/Koenkk/zigbee-herdsman-converters/pull/3339)
> 
> Make sure to first force remove your curtain switch from your zigbee network and pair it after you apply the changes from the pull request. Otherwise, it will still be recognized as the incorrect switch (double switch). The log should explicitly say curtain switch.
> 
> Good luck! Let me know if it works for you.

Thank you for your work, unfortunately when using the conbee II it is not possible to see the status of any livolo device.
I think it is due to a conbee II firmware problem, as I have read it is already reported, so for now I can not do much with livolo.  https://github.com/Koenkk/zigbee2mqtt/issues/6555 Regards.

**Comment by m3vip2** [comment]
I'm interested in this curtain switch, did you @Piscator74 stopped to research about controlling it?

**Comment by Piscator74** [comment]
I finally had some time to look at the [livolo curtain switch](https://www.zigbee2mqtt.io/devices/TI0001-curtain-switch.html) and to make it controllable. Basically, I tried payloads sent by [livolo roller blind motor](https://www.zigbee2mqtt.io/devices/TI0001-cover.html) and I had some luck.

I can now set the curtain switch left and right channel on for a fixed duration. However, I cannot control the direction or stop it once it is turned on. To try it out, add the code below to "toZigbee.js" and point to the "livolo_curtain_switch_on_off" function from 'TI0001-curtain-switch' in "livolo.js".

If you send the payload with 0x0401, the following happens:
* First time, left channel is activated for a duration specified by the first value.
* Second time, nothing happens.
* Third time, left channel is activated for a duration specified by the first value.
* Fourth time, nothing happens.
* And the sequence repeats.
In the code below, I send the payload twice. The control now works as a toggle between the left and the right button.

If anyone has a good idea on how to control the direction, please share.
(I obviously already tried the payloads from other functions such as "livolo_cover_state" and "livolo_cover_options")

```
    livolo_curtain_switch_on_off: {
    // debug notes:
    //    sending payload1 = { 0x0401: { value: [0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00], type: 1 } };
    //    result is: right = ON, but the switch hangs until repowered.
        key: ['state'],
        convertSet: async (entity, key, value, meta) => {
            if (typeof value !== 'string') {
                return;
            }
            const state = value.toLowerCase();
            const postfix = meta.endpoint_name || 'left';
            utils.assertEndpoint(entity);
            await entity.command('genOnOff', 'toggle', {}, { transactionSequenceNumber: 0 });
            let payload1;
            let payload2;
            const options = {
                manufacturerCode: 0x1ad2, disableDefaultResponse: true, disableResponse: true,
                reservedBits: 3, direction: 1, transactionSequenceNumber: 0xe9, writeUndiv: true,
            };  
            if ((postfix === 'left') && (state === 'on')) {
                meta.logger.debug('left on')
                payload1 = { 0x0401: { value: [0x10, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00], type: 1 } };
                payload2 = { 0x0401: { value: [0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00], type: 1 } };
            }
            else if ((postfix === 'right') && (state === 'on')) {
                meta.logger.debug('right on')
                payload1 = { 0x0401: { value: [0x10, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00], type: 1 } };
                payload2 = { 0x0401: { value: [0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00], type: 1 } };
            }
            else {
                meta.logger.debug('both off')
                payload1 = { 0x0401: { value: [0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00], type: 1 } };
                payload2 = { 0x0401: { value: [0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00], type: 1 } };
            }
            await entity.write('genPowerCfg', payload1, options);
            await entity.write('genPowerCfg', payload2, options);
            return { state: { state: value.toUpperCase() }, readAfterWriteTime: 250 };
        },
    },
```
